Moving PRF to wolfcrypt

embhorn opened the pull request:
Migrate the pseudo random function component to the wolfcrypt library. Also created a compile time define that can be used to enable PRF, `WOLFSSL_HAVE_PRF`.

embhorn commented on src/tls.c:
```diff
@@ -459,10 +174,15 @@ int BuildTlsFinished(WOLFSSL* ssl, Hashes* hashes, const byte* sender)
                    IsAtLeastTLSv1_2(ssl), ssl->specs.mac_algorithm,
                    ssl->heap, ssl->devId);
+#else
+        /* Pseudo random function must be enabled */
```
Got it.

embhorn commented:
Retest this please.

embhorn commented:
retest this please

embhorn commented:
retest this please

embhorn commented:
retest this please

toddouska commented on src/tls.c:
```diff
@@ -555,8 +278,27 @@ static int _DeriveTlsKeys(byte* key_dig, word32 key_dig_len,
                seed, SEED_LEN, tls1_2, hash_type, heap, devId);
+#else
+    /* Pseudo random function must be enabled */
+    ret = WOLFSSL_FATAL_ERROR;
```
We need a message here and shouldn't the return code be about not compile in?

toddouska commented on src/tls.c:
```diff
@@ -628,8 +370,21 @@ static int _MakeTlsMasterSecret(byte* ms, word32 msLen,
                seed, SEED_LEN, tls1_2, hash_type, heap, devId);
+#else
+    /* Pseudo random function must be enabled */
+    ret = WOLFSSL_FATAL_ERROR;
```
too generic, no message.

toddouska commented on src/tls.c:
```diff
@@ -657,8 +412,26 @@ static int _MakeTlsExtendedMasterSecret(byte* ms, word32 msLen,
                sHash, sHashLen, tls1_2, hash_type, heap, devId);
+#else
+    /* Pseudo random function must be enabled */
+    ret = WOLFSSL_FATAL_ERROR;
```
too generic, no message.

toddouska commented on src/tls.c:
```diff
@@ -786,10 +559,18 @@ int wolfSSL_make_eap_keys(WOLFSSL* ssl, void* msk, unsigned int len,
               ssl->heap, ssl->devId);
+#else
+    /* Pseudo random function must be enabled */
+    ret = WOLFSSL_FATAL_ERROR;
```
too generic, no message.

toddouska commented on wolfcrypt/src/hash.c:
```diff
@@ -928,3 +923,275 @@ int wc_HashFree(wc_HashAlg* hash, enum wc_HashType type)
+#endif
+
+/* compute p_hash for MD5, SHA-1, SHA-256, or SHA-384 */
+int wc_PRF(byte* result, word32 resLen, const byte* secret,
```
comment mentions p_hash but the name has changed.  This should be more consistent for clarity.

toddouska commented on wolfcrypt/src/hash.c:
```diff
@@ -928,3 +923,275 @@ int wc_HashFree(wc_HashAlg* hash, enum wc_HashType type)
+}
+
+/* Wrapper for TLS 1.2 and TLSv1 cases to calculate PRF */
+/* In TLS 1.2 case call straight thru to p_hash */
```
p_hash no longer exists.

toddouska commented on wolfcrypt/src/hash.c:
```diff
@@ -928,3 +923,275 @@ int wc_HashFree(wc_HashAlg* hash, enum wc_HashType type)
+         * should use better. */
+        if (hash_type < sha256_mac || hash_type == blake2b_mac)
+            hash_type = sha256_mac;
+        /* compute p_hash for MD5, SHA-1, SHA-256, or SHA-384 for TLSv1 PRF */
```
TLSv1 is wrong here.

toddouska replied:
Also, p_hash no longer exists.

embhorn commented:
Hi @toddouska 
I updated the comments that referenced "p_hash" and added a new error for the missing PRF condition, and a message in `wolfSSL_ERR_reason_error_string()`.

embhorn commented on wolfcrypt/src/hash.c:
```diff
@@ -1176,7 +1176,7 @@ int wc_PRF_TLS(byte* digest, word32 digLen, const byte* secret, word32 secLen,
         if (hash_type < sha256_mac || hash_type == blake2b_mac)
             hash_type = sha256_mac;
-        /* compute p_hash for MD5, SHA-1, SHA-256, or SHA-384 for TLSv1 PRF */
+        /* compute hash for MD5, SHA-1, SHA-256, or SHA-384 for TLSv1.2 PRF */
```
Got it.

embhorn commented on src/tls.c:
```diff
@@ -180,8 +180,8 @@ int BuildTlsFinished(WOLFSSL* ssl, Hashes* hashes, const byte* sender)
-        /* Pseudo random function must be enabled */
-        ret = WOLFSSL_FATAL_ERROR;
+        /* Pseudo random function must be enabled in the configuration. */
+        ret = PRF_MISSING;
```
Updated all instances.
